pkg/sources/adapter/salesforcesource/client/bayeux_test.go

```diff
@@ -0,0 +1,350 @@
+			responses: []response{
+				{handshake: handshakeResponse()},
+				{subscribe: subscribeResponse()},
+				{connect: connectResponse()},
+			},
+
+			expectedConnectResponses: 1,
+			expectedConnectErrors:    0,
+		},
+
+		"connect, receive 2 messages": {
+			responses: []response{
+				{handshake: handshakeResponse()},
+				{subscribe: subscribeResponse()},
+				{connect: connectResponse()},
+				{connect: connectResponse()},
+			},
+
+			expectedConnectResponses: 2,
+			expectedConnectErrors:    0,
+		},
+
+		"connect, with meta successful and 2 messages": {
+			responses: []response{
+				{handshake: handshakeResponse()},
+				{subscribe: subscribeResponse()},
+				{connect: connectResponse()},
+				{connect: connectResponse(connectWithChannel(tMetaChannel))},
+				{connect: connectResponse()},
+			},
+
+			expectedConnectResponses: 2,
+			expectedConnectErrors:    0,
+		},
+
+		"connect, with meta advicing handshake": {
+			responses: []response{
+				{handshake: handshakeResponse()},
+				{subscribe: subscribeResponse()},
+				{connect: connectResponse(
+					connectWithChannel(tMetaChannel),
+					connectWithSuccessful(false),
+					connectWithAdviceReconnect(tReconnect),
+				)},
+				{handshake: handshakeResponse()},
+				{subscribe: subscribeResponse()},
+				{connect: connectResponse()},
+			},
+
+			expectedConnectResponses: 1,
+			expectedConnectErrors:    0,
+		},
+
+		"connect, with meta advicing handshake and 2 messages": {
+			responses: []response{
+				{handshake: handshakeResponse()},
+				{subscribe: subscribeResponse()},
+				{connect: connectResponse()},
+				{connect: connectResponse(
+					connectWithChannel(tMetaChannel),
+					connectWithSuccessful(false),
+					connectWithAdviceReconnect(tReconnect),
+				)},
+				{handshake: handshakeResponse()},
+				{subscribe: subscribeResponse()},
+				{connect: connectResponse()},
+			},
+
+			expectedConnectResponses: 2,
+			expectedConnectErrors:    0,
+		},
+	}
+
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+
+			sf := httptest.NewServer(mockBayeuxServerHandler(tc.responses))
+			defer sf.Close()
+
+			dispatcher := &eventDispatcher{
+				eof: make(chan struct{}),
+			}
+			authenticator := fake.NewFakeAuthenticator(auth.Credentials{
+				InstanceURL: sf.URL,
+			})
+			ctx, cancel := context.WithCancel(context.Background())
+
+			b := NewBayeux(tAPIVersion, tSubscription, authenticator, dispatcher, sf.Client(), logger)
+
+			go func() {
+				clientErr := b.Start(ctx)
+				require.Nil(t, clientErr, "The bayeux client failed")
+			}()
+
+			select {
+			case <-time.After(4000 * time.Millisecond):
+				t.Fatal("Test timed out.")
+
+			case <-dispatcher.eof:
+			}
+
+			require.Equal(t, tc.expectedConnectResponses, len(dispatcher.dispatchedEvents), "Unexpected number of responses from connect")
+			require.Equal(t, tc.expectedConnectErrors, len(dispatcher.dispatchedErrors), "Unexpected number of errors from dispatcher")
+
+			cancel()
```

Comment:
This could be `defer`'d or `t.Cleanup`'d I believe.
Both work as expected inside a `t.Run` subtest.

Reply:
Good catch.
I did not know `t.Cleanup` even existed, using it at https://github.com/triggermesh/triggermesh/pull/352/commits/a995ac74b170e05e055731a2246e56e60508e49e instead of `defer` just for practice.